baseline/preprocessor/chunking_service.py

```diff
@@ -1 +1,243 @@
+    Chunking strategy that splits text into paragraphs.
+    """
+
+    def __init__(self, chunk_size: int = 1000):
+        """
+        Initialize the paragraph-based chunking strategy.
+
+        Args:
+            chunk_size (int): The size of each chunk in characters.
+        """
+        self.chunk_size = chunk_size
+
+    def chunk(self, text: str) -> list:
+        """
+        Chunk the text into paragraphs.
+
+        Args:
+            text (str): The text to be chunked.
+
+        Returns:
+            list: A list of text chunks.
+        """
+        paragraphs = text.split("\n\n")
+        chunks = []
+        current_chunk = ""
+
+        for paragraph in paragraphs:
+            if len(current_chunk) + len(paragraph) + 1 <= self.chunk_size:
+                current_chunk += ("\n\n" if current_chunk else "") + paragraph
+            else:
+                chunks.append(current_chunk)
+                current_chunk = paragraph
+
+        if current_chunk:
+            chunks.append(current_chunk)
+
+        return chunks
+
+
+class SemanticChunkingStrategy(ChunkingStrategy):
+    """
+    Chunking strategy that uses a recursive approach to split text into smaller pieces.
+    """
+
+    def __init__(self, chunk_size: int = 1000, chunk_overlap: int = 200):
+        """
+        Initialize the recursive chunking strategy.
+
+        Args:
+            chunk_size (int): The size of each chunk in characters.
+            chunk_overlap (int): The number of overlapping characters between chunks.
+        """
+        self.chunk_size = chunk_size
+        self.chunk_overlap = chunk_overlap
+
+    def chunk(self, text: str) -> list:
+        """
+        Chunk the text using a recursive approach.
+
+        Args:
+            text (str): The text to be chunked.
+
+        Returns:
+            list: A list of text chunks.
+        """
+        splitter = RecursiveCharacterTextSplitter(
+            chunk_size=self.chunk_size, chunk_overlap=self.chunk_overlap
+        ).split_text(text)
+
+        return splitter
+
+
+class MarkdownHeaderChunkingStrategy(ChunkingStrategy):
+    """
+    Chunking strategy that splits Markdown text based on header structure.
+    """
+    def __init__(self, headers_to_split_on: List[Tuple[str, str]] = None):
+        """
+        Initialize the markdown header chunking strategy.
+
+        Args:
+            headers_to_split_on (List[Tuple[str, str]]): A list of tuples specifying
+                the markdown headers to split on and their corresponding metadata keys.
+                Example: [("#", "Header1"), ("##", "Header2")]
+        """
+        if headers_to_split_on is None:
+            headers_to_split_on = [
+                ("#", "Header1"),
+                ("##", "Header2"),
+                ("###", "Header3"),
+                ("####", "Header4"),
+                ("#####", "Header5"),
+                ("######", "Header6"),
+            ]
+        self.splitter = MarkdownHeaderTextSplitter(
```

Comment:
1. does this splitter needs to be a class instance variable? I mean it will be same for all the instances and will be occupying more memory than it needs. Maybe it can be static as there is not instance specific need of this.
2. while making any private method or private variable, it's good to use `__<method/variable>